File: app/src/main/java/com/team2052/frckrawler/gui/MetricWidget.java
package com.team2052.frckrawler.gui;

import android.content.Context;
import android.view.LayoutInflater;
import android.widget.FrameLayout;

import com.team2052.frckrawler.database.DBContract;
import com.team2052.frckrawler.database.structures.Metric;
import com.team2052.frckrawler.database.structures.MetricValue;
import com.team2052.frckrawler.database.structures.MetricValue.MetricTypeMismatchException;

public abstract class MetricWidget extends FrameLayout {

    protected LayoutInflater inflater;
    private Metric metric;

    protected MetricWidget(Context context, Metric m, String[] val) {
        super(context);
        metric = m;
        inflater = (LayoutInflater) context.getSystemService
                (Context.LAYOUT_INFLATER_SERVICE);
    }

    public static MetricWidget createWidget(Context c, Metric m) {
        try {
            return createWidget(c, new MetricValue(m, new String[0]));
        } catch (MetricTypeMismatchException e) {
            return null;
        }
    }

    public static MetricWidget createWidget(Context c, MetricValue m) {
        if (m == null)
            return null;

        switch (m.getMetric().getType()) {
            case DBContract.BOOLEAN:
                return new BooleanMetricWidget(c, m);

            case DBContract.CHOOSER:
                return new ChooserMetricWidget(c, m);

            case DBContract.COUNTER:
                return new CounterMetricWidget(c, m);

            case DBContract.TEXT:
                return new TextMetricWidget(c, m);

            case DBContract.SLIDER:
                return new SliderMetricWidget(c, m);

            case DBContract.MATH:
                return new MathMetricWidget(c, m);

            default:

                return null;
        }
    }

    public Metric getMetric() {
        return metric;
    }

    public MetricValue getMetricValue() {

        try {
            return new MetricValue(metric, getValues());
        } catch (MetricTypeMismatchException e) {
            e.printStackTrace();
            return null;
        }
    }

    public abstract String[] getValues();
}

File: app/src/main/java/com/team2052/frckrawler/metric/view/MetricWidget.kt
package com.team2052.frckrawler.metric.view

import android.content.Context
import android.view.LayoutInflater
import android.widget.FrameLayout

import com.google.gson.JsonElement
import com.team2052.frckrawler.metric.data.MetricValue
import com.team2052.frckrawler.models.Metric

abstract class MetricWidget protected constructor(context: Context) : FrameLayout(context) {

    protected var inflater: LayoutInflater = context.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater
    lateinit var metric: Metric

    protected constructor(context: Context, m: MetricValue) : this(context) {
        metric = m.metric
        setMetricValue(m)
    }

    init {
        initViews()
    }

    abstract fun setMetricValue(m: MetricValue)
    abstract fun initViews()

    fun getValue(): MetricValue = MetricValue(metric, data)

    abstract val data: JsonElement
}
